### PR TITLE
Add MongoDB and NeDB Providers

### PR DESCRIPTION
**Pull request by CyberiumShadow**

### CODE REVIEW

File: providers/mongodb.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
+/* eslint-disable no-shadow */
+let db;
+const config = {
+  moduleName: "mongo",
+  enabled: true,
+  dbName: "Komada",
+  dbURL: "mongodb://localhost:27017/",
+};
+
+function connect() { db.connect(`${config.dbURL}${config.dbName}`); }
```
Comment (kyranet) [inline providers/mongodb.js]: This returns `void`, `connect().then()` will throw an error.

File: providers/mongodb.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+/**
+ * Drops a collection within a DB.
+ *
+ * @param {string} name
+ * @returns {Promise}
+ */
+exports.dropCollection = name => connect().then(db => db.dropCollection(name));
+
+// Document Methods. Update requires MongoDB Update Operators. Be sure to refer to MongoDB documentation.
```
Comment (kyranet) [inline providers/mongodb.js]: That comment must be inside the JSDoc.

File: providers/mongodb.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+/**
+ * 
+ *
+ * @param {string} collection
+ */
+exports.all = collection => connect().then(db => db.collection(collection).find({}).toArray());
+
+/**
+ *
+ *
```
Comment (kyranet) [inline providers/mongodb.js]: And please, avoid unnecessary multiple lines

File: providers/mongodb.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+ * @param {string} collection
+ * @param {string} id
+ */
+exports.get = (collection, id) => connect().then(db => db.collection(collection).findOne({ "id": id })); // eslint-disable-line quote-props
+
+/**
+ *
+ *
```
Comment (kyranet) [inline providers/mongodb.js]: Same here, and in the rest of JSDocs.

File: providers/mongodb.js [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+function connect() { db.connect(`${config.dbURL}${config.dbName}`); }
+
+// Collection Methods. Collections are implicitly created with document methods regardess.
+
+/**
+ * Create a collection within a DB. Options may be specfied, refer to MongoDB docs.
+ *
+ * @param {string} name Name of the Collection to crate
+ * @param {Object} [options] Object containing various options for the created Collection
+ * @returns {Promise} Returns a promise containing the created Collection.
```
Comment (kyranet) [inline providers/mongodb.js]: This returns a `Promise`, yeah, but what does that Promise return after being resolved? If it returns nothing, `Promise<void>`.

File: providers/mongodb.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+ */
+exports.createCollection = (name, options = {}) => connect().then(db => db.createCollection(name, options));
+
+/**
+ * Drops a collection within a DB.
+ *
+ * @param {string} name
```
Comment (kyranet) [inline providers/mongodb.js]: You must write a description for the parameter `name`. Same for the rest of JSDocs.

File: providers/mongodb.js [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+exports.replace = (collection, filter, repDoc) => connect().then(db => db.collection(collection).replaceOne(filter, repDoc));
+
+exports.eval = () => connect();
+
+exports.help = {};
+exports.help.name = "mongodb";
+exports.help.type = "providers";
+exports.help.description = "Allows you use MongoDB functionality throughout Komada.";
+exports.conf = config;
+exports.conf.requiredModules = ["mongodb"];
```
Comment (kyranet) [inline providers/mongodb.js]: You can use

```js
exports.help = {
  name: "mongodb",
  type: "providers",
  description: "Allows you to use MongoDB functionality throught Komada.",
};
```

Same for the exports.conf.

File: providers/mongodb.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+exports.init = () => {
+  db = require("mongodb").MongoClient;
+};
```
Comment (kyranet) [inline providers/mongodb.js]: Download v2 is able to download pieces with `require` outside module exports, take that advantage.

File: providers/nedb-class.js [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    this.db = null;
+  }
+
+  /**
+   * Connects to the NeDB datastore and populates this.db
+   *
+   * @param {?Client} client
+   * @returns {Datastore}
+   *
+   * @memberof NeDB
```
Comment (kyranet) [inline providers/nedb-class.js]: This is not neccessary. Same for the rest of JSDocs.

File: providers/nedb.js [hunk 10]
```diff
@@ -0,0 +1,82 @@
+let db;
+let Datastore;
+
+const config = {
+  persistent: false,
+};
+
+exports.init = (client) => {
```
Comment (kyranet) [inline providers/nedb.js]: You took the advantage of Download v2 being able to recognize `require` outside module exports in the nedb-class provider.

File: providers/nedb.js [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+ *
+ * @param {Object} query
+ * @param {boolean} [all=false]
+ */
+exports.delete = (query, all = false) => db.removeAsync(query, { multi: all });
+
+/**
+ *
+ *
+ */
```
Comment (kyranet) [inline providers/nedb.js]: Unfinished JSDocs. Same for the JSDocs below.

**Review by kyranet — DISMISSED** [review]
There are your requested changes.

File: providers/mongodb.js [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+ */
+exports.all = collection => db.collection(collection).find({}).toArray();
+
+/**
+ * Retrieves a single Document from a Collection that matches a user determined ID
+ * @param {string} collection Name of the Collection
+ * @param {string} id ID of the document
+ * @returns {Promise<?Object>}
+ */
+exports.get = (collection, id) => db.collection(collection).findOne({ "id": id }); // eslint-disable-line quote-props
```
Comment (kyranet) [inline providers/mongodb.js]: `findOne({ id })` is perfectly fine.

File: providers/mongodb.js [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+  db = await Mongo.connect(`${config.dbURL}${config.dbName}`);
+};
+
+// Collection Methods. Collections are implicitly created with document methods regardess.
+
+/**
+ * Create a collection within a DB. Options may be specfied, refer to MongoDB docs.
+ * @param {string} name Name of the Collection to crate
+ * @param {Object} [options] Object containing various options for the created Collection
+ * @returns {Promise<Collection>} Returns a promise containing the created Collection.
```
Comment (kyranet) [inline providers/mongodb.js:28]: Yes, but what is `Collection`? A class? Or simply an Object?

File: providers/mongodb.js [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+ * @param {string} id ID of the document
+ * @returns {Promise<?Object>}
+ */
+exports.get = (collection, id) => db.collection(collection).findOne({ "id": id }); // eslint-disable-line quote-props
+
+/**
+ * Inserts a Document into a Collection using a user provided object.
+ * @param {string} collection Name of the Collection
+ * @param {Object} docObj Document Object to insert
+ * @returns {Promise<CommandResult>}
```
Comment (kyranet) [inline providers/mongodb.js]: Again, what's `CommandResult`?

File: providers/mongodb.js [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+ */
+exports.add = (collection, docObj) => db.collection(collection).insertOne(docObj);
+
+/**
+ * Deletes a Document from a Collection that matches a user determined ID *
+ * @param {string} collection Name of the Collection
+ * @param {string} id ID of the document
+ * @returns {Promise<CommandResult>}
```
Comment (kyranet) [inline providers/mongodb.js:88]: Same here.

File: providers/mongodb.js [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+ */
+exports.delete = (collection, id) => db.collection(collection).deleteOne({ id });
+
+/**
+ * Updates a Document using MongoDB Update Operators. *
+ * @param {string} collection Name of the Collection
+ * @param {Object} filter The Filter used to select the document to update
+ * @param {Object} updateObj The update operations to be applied to the document
+ * @returns {Promise<CommandResult>}
```
Comment (kyranet) [inline providers/mongodb.js]: Here.

File: providers/mongodb.js [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+ */
+exports.update = (collection, filter, updateObj) => db.collection(collection).updateOne(filter, updateObj);
+
+/**
+ * Replaces a Document with a new Document specified by the user *
+ * @param {string} collection Name of the Collection
+ * @param {Object} filter The Filter used to select the document to update
+ * @param {Object} repDoc The Document that replaces the matching document
+ * @returns {Promise<CommandResult>}
```
Comment (kyranet) [inline providers/mongodb.js:108]: And here.

File: providers/nedb-class.js [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+    this.db = null;
+  }
+
+  /**
+   * Connects to the NeDB datastore and populates this.db
+   *
+   * @param {?Client} client Komada Client
+   * @returns {Datastore} The NeDB Datastore itself
+   *
+   * @memberof NeDB
```
Comment (kyranet) [inline providers/nedb-class.js]: The memberof is not required. And you can remove the space that is between the description and param. Same for the rest of the file.

File: providers/nedb-class.js [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+   */
+  connect(client) {
+    return this.db = new Datastore(this.persistent ? { filename: resolve(`${client.clientBaseDir}/bwd/db/${this.name}-nedb.db`), autoload: true } : {}); // eslint-disable-line no-return-assign
```
Comment (kyranet) [inline providers/nedb-class.js]: Use `resolve(...pathPieces)` like we're doing now in any piece.

```js
resolve(client.clientBaseDir, "bwd", "provider", "nedb");
```

File: providers/nedb.js [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+const config = {
+  persistent: false,
+};
+
+/**
+ * Initiates the NeDB Provider
+ * @param {?Client} client Komada Client
+ */
+exports.init = (client) => {
+  db = new Datastore(config.persistent ? { filename: resolve(`${client.clientBaseDir}/bwd/db/komada-nedb.db`), autoload: true } : {});
```
Comment (kyranet) [inline providers/nedb.js]: Same here. (The resolve suggestion)

File: providers/nedb.js [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+exports.update = (query, updateDoc) => db.updateAsync(query, updateDoc);
+
+/**
+ * Deletes documents matching the query
+ *
+ * @param {any} query Query Object for Property Equality or using Query Operators
+ * @param {boolean} [all=false] Delete All
+ * @returns {Promise<number>} Returns the numbers of Documents deleted
+ */
+exports.delete = (query, all = false) => db.removeAsync(query, { multi: all });
```
Comment (kyranet) [inline providers/nedb.js]: This is incompatible with SettingGateway. Isn't there something like tables? SG requires the `delete` method to have the following parameters:

| Param | Type | Description |
| :---: | ---- | ----------- |
| table | `String` | The table name. |
|  row  | `String` | The ID of the document to delete |

**Review by kyranet — DISMISSED** [review]
Some requested changes.

File: providers/nedb.js (hunk 21 above)
Comment (CyberiumShadow) [reply providers/nedb.js]: No. NeDB is a table upon itself

**Comment by CyberiumShadow** [comment]
@kyranet can you please confirm that the NeDB and the MongoDB providers are now SG-Compliant and can handle async methods

**Comment by CyberiumShadow** [comment]
@kyranet pls review

File: providers/mongodb.js [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+
+/**
+ * Retrieves a single Document from a Collection that matches a user determined ID
+ * @param {string} collection Name of the Collection
+ * @param {string} id ID of the document
+ * @returns {Promise<?Object>}
+ */
+exports.get = (collection, id) => db.collection(collection).findOne(resolveQuery(id)); // eslint-disable-line quote-props
+
+exports.has = async (...args) => {
```
Comment (kyranet) [inline providers/mongodb.js:62]: Needs documentation for Provider#has

File: providers/mongodb.js [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+  const results = await this.get(...args);
+  if (!results) return false;
+  return true;
+};
+
+exports.getRandom = async (...args) => {
```
Comment (kyranet) [inline providers/mongodb.js:68]: Needs documentation for Provider#getRandom

File: providers/mongodb.js [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+/**
+ * Replaces a Document with a new Document specified by the user *
+ * @param {string} collection Name of the Collection
+ * @param {Object} filter The Filter used to select the document to update
+ * @param {Object} repDoc The Document that replaces the matching document
+ * @returns {Promise<CommandResult>}
+ */
+exports.replace = (collection, filter, repDoc) => db.collection(collection).replaceOne(resolveQuery(filter), repDoc);
+
+exports.eval = () => db;
```
Comment (kyranet) [inline providers/mongodb.js:112]: Needs documentation for Provider#eval

File: providers/nedb.js [hunk 25]
```diff
@@ -0,0 +1,153 @@
...
+ * @param {Object} data The data you want to update.
+ * @returns {Promise<number>} Returns a Promise containing the number of Documents Updated. (Either 0 or 1).
+ */
+exports.update = async (table, query, data) => {
+  const res = await this.get(table, query);
+  await dataStores.get(table).updateAsync(resolveQuery(query), Object.assign(res, data));
+  await dataStores.get(table).persistence.compactDatafile();
+  return true;
+};
+exports.replace = async (table, query, data) => {
```
Comment (kyranet) [inline providers/nedb.js:113]: Needs documentation for Provider#replace

**Review by kyranet — CHANGES_REQUESTED** [review]
You missed some documentation, and fix these spaces in the RethinkDB JSDocs lol

**Comment by kyranet** [comment]
You left 2 reviews without solve in regards to Collection and CommandStore, JSDocs has no idea of what are them, and technically they return objects, not classes.

**Review by kyranet — APPROVED** [review]

**Pull request merged**